src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,192 @@
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                throw socketOpenExceptions.OfType<SocketException>().FirstOrDefault()
+                      ?? (Exception)new AggregateException(socketOpenExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
+        }
+
+        private async ValueTask<Socket> OpenSocket(IPEndPoint ipEndPoint, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            var socket = new Socket(ipEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false };
+
+            // Enable keep-alive
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, KeepAliveIntervalSeconds);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, KeepAliveTimeSeconds);
+
+            if (isAsync)
+            {
+                #if NET6_0_OR_GREATER
+                await socket.ConnectAsync(ipEndPoint, ct).ConfigureAwait(false);
+                #else
+                await new TaskFactory(ct).FromAsync(socket.BeginConnect, socket.EndConnect, ipEndpoint, null)
+                    .ConfigureAwait(false);
+                #endif
+            }
+            else
+            {
+                OpenSocketSync(socket, ipEndPoint, ct);
+            }
+
+            // Connection is established
+            socket.Blocking = true;
+            socket.NoDelay = true;
+
+            return socket;
+        }
+
+        private void OpenSocketSync(Socket socket, IPEndPoint ipEndPoint, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            try
+            {
+                socket.Connect(ipEndPoint);
+            }
+            catch (SocketException e)
+            {
+                // Because the socket is configured to be non-blocking, any operation that would
+                // block will throw an exception indicating it would block. Since opening a TCP
+                // connection will always block, we expect to get an exception for it, and will
+                // ignore it. This allows us to immediately return from connect and poll it,
+                // allowing us to observe timeouts and cancellation.
+                if (e.SocketErrorCode is not SocketError.WouldBlock)
+                {
+                    throw;
+                }
+            }
+
+            // Poll the socket until it is open
+            if (!socket.Poll(DefaultPollTimeout, SelectMode.SelectWrite))
```

Comment:
Perhaps worth reducing the default poll timeout to something aggressive like 1/2/5/10/20/50ms and looping; this'll provide you with more opportunity to call `ct.ThrowIfCancellationRequested`. This poll timeout will be a tradeoff between cancellation precision and CPU usage, I've got no preference or data to suggest what the right figure for this is.

Reply:
I've thought about this as well. I'll add a todo about it to address when we really tackle cancellation.